### ISSUE TITLE
[Discussion] Issues related to Float

### ISSUE THREAD

**Issue by czgdp1807** [sympy/sympy#16918]
#### Topics of discussion

1. How should precision affect the result?
2. Should Floats like `1.0` be treated as integers? Technically, is `Float(1.0).is_integer == True` a valid consideration?

#### Other comments/references
[1] https://github.com/sympy/sympy/pull/16897
[2] https://github.com/sympy/sympy/pull/16916
There are many comments over other issues but since, they are distributed, definite conclusions are not reached on the above topics.
ping @smichr @jksuom @oscarbenjamin

**Comment by oscarbenjamin** [comment]
I think the fundamental problem in SymPy is that the concepts of *accuracy* and *precision* are conflated. These words have different meanings in different contexts and are not always used consistently even by numericists. I will give definitions here:

* Floating point: A floating point number is a number of the form `m*b**e` where `m` is the mantissa, `e` is the exponent and `b` is the base. Most commonly `b` is 2 or 10 and for SymPy's *binary* Floats it is 2. This means that SymPy's floats can exactly represent only rational numbers of the form `p/q` where `q` is a power of 2.

* Precision: The precision of a floating point number refers to the number of digits used for the mantissa `m`. Binary floating point numbers with mantissa `m` and a precision of `n` binary digits can exactly represent rational numbers `p/q` where `q` is a power of 2 and `-2**n <= p <= 2**n`.

* Accuracy: If the Float x is viewed as an approximation of some true value xtrue then the *error* in x is `x - xtrue`. The accuracy (where x and xtrue are nonzero) can be quantified in binary digits as `log2(abs(x / (x - xtrue)))`.

Most multi-*precision* libraries allow control of precision but leave accuracy as a problem for the user. That is to say that the library allows you to say "use a precision of 100 digits" and then accurately computes *individual* calculations using that level of precision. The library however does not guarantee that the end result of a complicated sequence of calculations has any particular *accuracy* compared to exact computations with the same inputs.

There are two broad issues that I see coming up repeatedly with Floats in SymPy. One is a problem that bothers users and the other is a problem that bothers more experienced contributors.

1. A lot of users want to be able to write 0.1 and have it mean precisely `1/10`. They find it more convenient to write numbers in decimal than rational form and are bothered by the fact that doing so can completely change the output of their calculations, can mean that integrals don't compute, polynomials don't find roots, algebraic manipulations don't work etc.

2. SymPy's concept of "precision" confuses accuracy and precision (as defined above) and means different things in different places. The result of an evalf call has an associated "precision" that is supposed to represent *accuracy*. Viewing the precision as representing accuracy leads to the desire to treat a floating point number as a range of numbers: hence `Float(1).is_integer -> False` (or perhaps should be None). However it isn't clear that it is always possible to treat accuracy like this when doing ordinary arithmetic in which every operation needs an associated precision but there isn't always an available object representing the "true" answer.

Problem 1. could be easily solvable if SymPy didn't need to interact with Python floats. There could be a decimal or decimal-rational type that has an exact rational representation and exactly converts input decimal strings. Under the hood it is treated as rational but it prints in decimal form. Division by numbers with prime factors other than 2 or 5 could create something like `0.1/3` or perhaps just a Rational that prints as `0.0333`. The problem is how to interact with Python's float types. If a user types 0.1 that creates a Python float which is already a binary float so `0.1` is already not exactly `1/10`. SymPy could use some sort of nsimplify here but that would have some chance of introducing small errors (maybe that's an acceptable price for float->Float conversions - users who care can call an explicit conversion function).

**Comment by asmeurer** [comment]
I remember we decided at one point that Float.is_rational should give None (if you search the history of the code you can probably find the discussion). is_integer should probably do the same.

**Comment by czgdp1807** [comment]
> Problem 1. could be easily solvable if SymPy didn't need to interact with Python floats. There could be a decimal or decimal-rational type that has an exact rational representation and exactly converts input decimal strings. Under the hood it is treated as rational but it prints in decimal form. Division by numbers with prime factors other than 2 or 5 could create something like `0.1/3` or perhaps just a Rational that prints as `0.0333`. The problem is how to interact with Python's float types. If a user types 0.1 that creates a Python float which is already a binary float so `0.1` is already not exactly `1/10`. SymPy could use some sort of nsimplify here but that would have some chance of introducing small errors (maybe that's an acceptable price for float->Float conversions - users who care can call an explicit conversion function).

Supportive of this method.
@asmeurer Let's try the change `Float.is_integer` from `False` -> `None` and then we will see how does it affect the things in SymPy.

**Comment by sylee957** [comment]
About the gamma functions, I think it can be resolved by using numeric comparisons to give complex infinity, without touching the assumptions system.

**Comment by czgdp1807** [comment]
I think rather than creating a new decimal-rational type we should, provide an extra attribute of `as_rational` to convert the decimal to rational and printing can also be done according to `as_rational`.
Another point to be noted is, that by default for every symbolic expression, any Float in that expression should be converted to Rational automatically.

**Comment by leogama** [comment]
Since `float('1').is_integer() == True` and `Float('1').is_rational == None`, there are just two sane (consistent) possibilities for me. It should be either
* `Float('1').is_integer` -> `True`
or 
* `Float('1').is_integer` -> `None`

In other words, the following seems absurd:
```python
a = 1
assert Float(a).is_integer == (not float(a).is_integer())  # passes
```

If the final decision were to make no changes at all, this behavior should be made explicit on documentation and tests at least. Good luck! :slightly_smiling_face:

**Comment by asmeurer** [comment]
The `is_integer` method on the Python built-in `float` is completely unrelated to the SymPy `is_integer` attribute. They happen to share the same name, but don't confuse them.

**Comment by asmeurer** [comment]
Although I don't quite understand this

```py
>>> Float(1).is_integer
False
>>> print(Float(1).is_rational)
None
```

Why don't we also use None for Float.is_integer?

**Comment by oscarbenjamin** [comment]
`Float(1).is_integer` should not give `False`

**Comment by czgdp1807** [comment]
`Float.is_integer` should give `None` for integer inputs.

**Comment by leogama** [comment]
> The `is_integer` method on the Python built-in `float` is completely unrelated to the SymPy `is_integer` attribute. They happen to share the same name, but don't confuse them.

Yes, I _understand_ these methods are not related at all. But, even if it was documented, the current behavior would still violate the "Principle of Least Astonishment" and inevitably confuse new users —even the `None` return value for `is_rational` and `is_integer` properties is somewhat confusing at first sight.

At the same time, the current definition for `Float.is_integer` is identical to the `Float.is_zero` one. So we would lose nothing by changing the former, and code relying on it could be updated smoothly.
